integration_test_suite/integration_tests/src/test/java/io/prometheus/jmx/test/httpserver/general/BlackListAndWhiteListObjectNames_IT.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2022-2023 Douglas Hoard
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.prometheus.jmx.test.httpserver.general;
+
+import io.prometheus.jmx.test.DockerImageNameParameters;
+import io.prometheus.jmx.test.HttpClient;
+import io.prometheus.jmx.test.Metric;
+import io.prometheus.jmx.test.MetricsParser;
+import io.prometheus.jmx.test.httpserver.BaseHttpServer_IT;
+import io.prometheus.jmx.test.support.ContentConsumer;
+import io.prometheus.jmx.test.support.HealthyRequest;
+import io.prometheus.jmx.test.support.HealthyResponse;
+import io.prometheus.jmx.test.support.MetricsRequest;
+import io.prometheus.jmx.test.support.MetricsResponse;
+import io.prometheus.jmx.test.support.OpenMetricsRequest;
+import io.prometheus.jmx.test.support.OpenMetricsResponse;
+import io.prometheus.jmx.test.support.PrometheusMetricsRequest;
+import io.prometheus.jmx.test.support.PrometheusMetricsResponse;
+import org.antublue.test.engine.api.Parameter;
+import org.antublue.test.engine.api.TestEngine;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.Network;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static io.prometheus.jmx.test.support.AssertThatRequestResponse.assertThatRequestResponse;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class BlackListAndWhiteListObjectNames_IT extends BaseHttpServer_IT implements ContentConsumer {
+
+    private static Network network;
+
+    private String dockerImageName;
+    private GenericContainer<?> applicationContainer;
+    private GenericContainer<?> exporterContainer;
+    private HttpClient httpClient;
+
+    @TestEngine.ParameterSupplier
+    public static Stream<Parameter> parameters() {
+        return DockerImageNameParameters.parameters();
+    }
+
+    @TestEngine.Parameter
+    public void parameter(Parameter parameter) {
+        dockerImageName = parameter.value();
+    }
+
+    @TestEngine.BeforeClass
+    public static void beforeClass() {
+        network = createNetwork();
+    }
+
+    @TestEngine.BeforeAll
+    public void beforeAll() {
+        applicationContainer = createApplicationContainer(this, dockerImageName, network);
+        applicationContainer.start();
+
+        exporterContainer = createExporterContainer(this, dockerImageName, network);
```

Comment:
You put a lot of work into reducing copy-and-paste, thanks a lot!

I'm wondering if we can reduce copy-and-paste even more. Most tests are duplicated: One for the `agent` mode, one for the `standalone` mode. We might be able to write just a single test, and add the mode as a parameter.

Antublue supports `ParameterMap` as a wrapper for multiple parameters. So instead of having a `Parameter` with just the Docker image name, we could use a `ParameterMap` containing the image name and the mode. For example, we could add something like the following to `DockerImageNameParameters.parameters()`:

```java
        for (String mode : new String[]{"standalone", "agent"}) {
            for (String dockerImageName : dockerImageNames) {
                Parameter parameter = ParameterMap
                        .named(dockerImageName + " " + mode)
                        .put("image", dockerImageName)
                        .put("mode", mode)
                        .parameter();
                if (predicate.test(parameter)) {
                    parameters.add(parameter);
                }
            }
        }
```

Then the `parameter()` method in the tests would look something like this:

```java
    @TestEngine.Parameter
    public void parameter(Parameter parameter) {
        ParameterMap parameterMap = parameter.value(ParameterMap.class);
        dockerImageName = parameterMap.get("image");
        mode = parameterMap.get("mode");
    }
```

and in the tests initialization code we could do something like this (the initialization code can be implemented in a common utility class or base class because it will be the same for most tests):

```java
        if (mode.equals("standalone")) {
            // run the standalone http server
            // ...
        } else {
            // run the java agent
            // ...
        }
```

What do you think?

Reply:
I refactored the tests, consolidating the Java agent and standalone tests. The code is much, much smaller, and cleaner.

The execution time has _very slightly_ increased since this approach decreases parallelism (the test engine executes a test class sequentially, but different test classes are executed in parallel.)